File: packages/core/src/engine/pension.ts
import type { YearData, Rounding } from '../types';
import { findBracket } from '../lookup/bracket';
import { applyRate, numberToRateString } from '../rounding/strategies';

export interface PensionResult {
  insured: number;
  employer: number;
  self: number;
}

/**
 * 勞退. Uses the pension month-contribution table (cap 150,000).
 *   employer = round(insured × 6%); self = round(insured × selfRate), selfRate ∈ [0, 0.06].
 */
export function calcPension(
  data: YearData,
  monthlySalary: number,
  selfContributionRate: number,
  rounding: Rounding,
): PensionResult {
  const p = data.pension;
  const insured = findBracket(p.brackets, monthlySalary).insuredSalary;
  const clampedRate = Math.min(Math.max(selfContributionRate, 0), 0.06);
  // numberToRateString uses toFixed(10) to avoid scientific notation (e.g. "1e-8") that
  // parseRate would reject; 10 decimals is finer than any real提繳率.
  const selfRate = numberToRateString(clampedRate);
  return {
    insured,
    employer: applyRate(insured, [p.employerRate], rounding),
    self: clampedRate > 0 ? applyRate(insured, [selfRate], rounding) : 0,
  };
}

File: packages/python/taiwan_payroll/engine/pension.py
from __future__ import annotations

from .._lookup import find_bracket
from .._rounding import apply_rate, number_to_rate_string


def calc_pension(data: dict, monthly_salary: float, self_contribution_rate: float, rounding: str) -> dict:
    p = data["pension"]
    insured = find_bracket(p["brackets"], monthly_salary)["insuredSalary"]
    clamped = min(max(self_contribution_rate, 0.0), 0.06)
    self_rate = number_to_rate_string(clamped)
    return {
        "insured": insured,
        "employer": apply_rate(insured, [p["employerRate"]], rounding),
        "self": apply_rate(insured, [self_rate], rounding) if clamped > 0 else 0,
    }
